Add the downscale functions that works for North America SDA outputs.

DongchenZ opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I agree that PEcAn Project may distribute my contribution under any or all of
	- the same license as the existing code,
	- and/or the BSD 3-clause license.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

Commit:
Update namespace for added functions.

Commit:
Added the script for running the north america downscale functions.

Commit:
Add the script for the downscale functions.

Commit:
Update the Rd files.

Commit:
Add function to namespace.

Commit:
Update function.

Commit:
Parallel average ERA5.

Commit:
Update documentation.

Commit:
Merge branch 'SDA_downscale' of https://github.com/DongchenZ/pecan into SDA_downscale

Commit:
Update change log.

Commit:
Remove commented lines.

mdietze commented on modules/assim.sequential/NAMESPACE:
```diff
@@ -1,6 +1,8 @@
 # Generated by roxygen2: do not edit by hand
 
+S3method(stack,covariates.2.geotiff)
```
I suspect this one isn't intended to be a S3method. What about renaming the function with underscores instead of periods?

DongchenZ replied:
Fixed.

mdietze commented on modules/assim.sequential/R/SDA_NA_downscale.R:
```diff
@@ -0,0 +1,459 @@
+      end <- as.Date(paste0(lubridate::year(dates[i]), "-12-31"))
+    }
+    # loop over years.
+    for (j in seq_along(dates)) {
```
not following why there's a nested loop over years within the loop oveer years, especially since I don't see where `j` is used

DongchenZ replied:
Fixed.

mdietze commented on modules/assim.sequential/R/SDA_NA_downscale.R:
```diff
@@ -0,0 +1,459 @@
+  VPD <- PEcAn.data.atmosphere::get.vpd(100*relative.humidity, temp)
+  # combine together.
+  PEcAn.logger::logger.info("Aggregate maps.")
+  met.rast <- c(terra::rast(matrix(temp, nrow = dim(temp)[2], ncol = dim(temp)[1], byrow = T)),
```
If I understand correctly, the goal seems to be to generate a single map of mean climate over a specific time period. If you wanted the downscaling to use year-specific meteorology, would you call this function in a loop, iteratively calculating from July 15 of one year to July 14 of the next year, or would it make more sense to have this function generate multiple met layers at different resolutions? Similarly, if/when the SDA drops to a monthly or seasonal timestep, could this function handle calculating monthly mean weather?

DongchenZ replied:
Fixed. The temporal resolution depends on the start and end dates.

mdietze commented on modules/assim.sequential/R/SDA_NA_downscale.R:
```diff
@@ -0,0 +1,459 @@
+
+#' @description
+#' This function helps to create the training dataset of specific variable type and locations for downscaling.
+#' TODO: There will be a ratio argument between training and testing samples to testify the ML regression accuracy.
```
not following what the TODO is about

DongchenZ replied:
Fixed.

mdietze commented on modules/assim.sequential/R/SDA_NA_downscale.R:
```diff
@@ -0,0 +1,459 @@
+#' @param covariates.dir character: path to the exported covariates GeoTIFF file.
+#' @param variable character: name of state variable. It should match up with the column names of the analysis data frame. 
+#'
+#' @return matrix within which the first sets of columns contain values of state variables for each ensemble mebers of every site, and the rest columns contain the corresponding covariates.
```
1. clarify the expected dimensions of the matrix to make how things are organized more clear. 2. fix typos

DongchenZ replied:
Fixed.

mdietze commented on modules/assim.sequential/R/SDA_NA_downscale.R:
```diff
@@ -0,0 +1,459 @@
+#' @author Dongchen Zhang
+prepare.train.dat <- function(settings, analysis, covariates.dir, variable) {
+  # convert settings into geospatial points.
+  pts <- pecan.settings.2.pts(settings)
```
If the settings object is only used to get the points, why not just have the points be the function argument? That gives you more flexibility to have pulled the points from the csv or shapefile

DongchenZ replied:
Fixed.

mdietze commented on modules/assim.sequential/R/SDA_NA_downscale.R:
```diff
@@ -0,0 +1,459 @@
+#' @description
```
filename includes "NA" which is confusing because 1. for most R users that means Not Available and 2. there's nothing North America specific to any of these functions. The NA specific bit is in the inst/anchor script

DongchenZ replied:
Fixed.

mdietze reviewed (changes requested):
So at a high level, these new fixes will create duplications between Josh's old code, @dlebauer new code, and @DongchenZ new code in this PR. I'm Ok with that in short term, but highly encourage @DongchenZ @dlebauer and Yang (who wants to downscale fluxes using XGBoost) to work together to come up with an agreed-upon set of functions, arguments, and workflows, and to get these into the bookdown documentation.  I also might suggest that assim.sequential might not be the best long-term home for this code, as it's not part of the SDA itself and can be used to downscale any ensemble runs, not just SDA -- maybe modules/data.mining (which itself could get an updated name, to something like modules/ML?)

Commit:
Rename the function.

Commit:
Remove the redundant for loop.

Commit:
Rename functions.

Commit:
Update documentation.

Commit:
Update the script for downscaling.

Commit:
Resolve conflicts.

Merge branch 'develop' of https://github.com/PecanProject/pecan into SDA_downscale

# Conflicts:
#	CHANGELOG.md

Commit:
Rename the file.

Commit:
Update documentation.

Commit:
Improve efficiency.

Commit:
Add varied land cover and stand age.

Commit:
Update documentation.

Commit:
Change the file path.

Commit:
Update path.

Commit:
change variable and path.

Commit:
Update path

Commit:
Update path.

Commit:
Update path.

Commit:
Update path.

mdietze commented:
@DongchenZ there are a few small errors blocking this PR:

```
 R check of modules/assim.sequential reports the following new problems. Please fix these and resubmit:
  checking dependencies in R code ... WARNING
  '::' or ':::' imports not declared from:
    ‘abind’ ‘doSNOW’ ‘foreach’ ‘itertools’ ‘PEcAn.data.atmosphere’
  checking Rd \usage sections ... WARNING
  Undocumented arguments in Rd file 'downscale_rf_main.Rd'
    ‘time’ ‘variable’
  Documented arguments not in \usage in Rd file 'downscale_rf_main.Rd':
    ‘time:’ ‘variable:’
  
  Undocumented arguments in Rd file 'parallel_prediction.Rd'
    ‘cov.vecs’
  Documented arguments not in \usage in Rd file 'parallel_prediction.Rd':
    ‘cov.vecs:’
```

Commit:
Update path

Commit:
Add xgboost to the description.

Commit:
Add the xgboost method to the downscale function and rename the previous RF functions to a more general name.

DongchenZ committed:
Merge branch 'PecanProject:develop' into SDA_downscale

Commit:
Merge branch 'SDA_downscale' of https://github.com/DongchenZ/pecan into SDA_downscale

Commit:
Update documentation.

Commit:
Update documentation.

Commit:
Update dependencies.

Commit:
Move the ERA5 averaging function to the atmosphere package.

Commit:
Update documentation and add package checking code.

Commit:
move the ERA5 average function to the atmosphere package.

Commit:
Update ERA5 average function namespace.

Commit:
Update documentation.

Commit:
Remove namespace.

Commit:
Update the document.

Commit:
Change package detection method.

Commit:
Update document.

Commit:
Add %dopar% namespace.

Commit:
Add a namespace for the variable inside the foreach.

Commit:
Update namespace.

Commit:
Update dependency.

Commit:
Remove package detection.
